library/core/registry/src/main/java/org/quiltmc/qsl/registry/api/QuiltRegistryBuilder.java

```diff
@@ -0,0 +1,311 @@
+	 * public class Thing {
+	 *     private final Holder<Thing> registryHolder = YourCoolModRegistries.THINGS.createIntrusiveHolder(this);
+	 *
+	 *     public Holder<Thing> getRegistryHolder() {
+	 *         return this.registryHolder;
+	 *     }
+	 * }
+	 * </code></pre>
+	 * Then, set this to call that {@code getRegistryHolder()} method:
+	 * <pre><code>
+	 * public class YourCoolModRegistries {
+	 *     public static final SimpleRegistry<Thing> THINGS =
+	 *         new QuiltRegistryBuilder(YourCoolMod.id("things"))
+	 *             .customHolderProvider(Thing::getRegistryHolder)
+	 *             .build();
+	 * }
+	 * </code></pre>
+	 * <p>
+	 * By default, this is {@code null} - the registry will create and maintain its own holders.
+	 *
+	 * @param customHolderProvider the new custom holder provider
+	 * @return this builder
+	 */
+	@Contract("_ -> this")
+	public @NotNull QuiltRegistryBuilder<T> customHolderProvider(@Nullable Function<T, Holder.Reference<T>> customHolderProvider) {
+		this.customHolderProvider = customHolderProvider;
+		return this;
+	}
+
+	/**
+	 * Sets the default identifier of this registry.
+	 * <p>
+	 * Should a nonexistent entry be referenced in some way, the registry will instead reference the entry identified
+	 * by this instead.
+	 * <p>
+	 * By default, this is {@code null} - the registry will simply return {@code null} when a nonexistent entry is referenced.
+	 *
+	 * @param defaultId the new default identifier
+	 * @return this builder
+	 */
+	@Contract("_ -> this")
+	public @NotNull QuiltRegistryBuilder<T> defaultId(@Nullable Identifier defaultId) {
+		this.defaultId = defaultId;
+		return this;
+	}
+
+	/**
+	 * Configures this builder to create a <em>built-in</em> registry
+	 * (that is registered in the {@linkplain Registry#REGISTRIES root registry}).
+	 * <p>
+	 * Built-in registry synchronizing is handled by us, and the synchronization behavior can be customized
+	 * via the {@link #syncBehavior(SyncBehavior)} method.
+	 * <p>
+	 * By default, this builder creates built-in registries.
+	 *
+	 * @return this builder
+	 */
+	@Contract("-> this")
+	public @NotNull QuiltRegistryBuilder<T> builtin() {
+		this.builtin = true;
+		return this;
+	}
+
+	/**
+	 * Configures this builder to create a <em>dynamic</em> registry
+	 * (that is <em>not</em> registered in the {@linkplain Registry#REGISTRIES root registry}).
+	 * <p>
+	 * Dynamic registry synchronizing is <em>not</em> handled by us.
+	 * <p>
+	 * Note that this method does <em>not</em> handle registering the registry into the
+	 * {@link net.minecraft.util.registry.DynamicRegistryManager DynamicRegistryManager}.
+	 * You will have to do this manually.
+	 *
+	 * @return this builder
+	 *
+	 * @throws IllegalArgumentException if this builder is configured with a {@linkplain #syncBehavior(SyncBehavior) synchronization behavior} that is not {@link SyncBehavior#SKIPPED}
+	 */
+	@Contract("-> this")
+	public @NotNull QuiltRegistryBuilder<T> dynamic() {
```

Comment:
So, I looked a bit into dynamic registries and to make this fully working we would need:
 - register the registry into [`BuiltinRegistries.REGISTRIES`](https://maven.quiltmc.org/repository/release/org/quiltmc/quilt-mappings/1.19.2+build.12/quilt-mappings-1.19.2+build.12-javadoc.jar/net/minecraft/util/registry/BuiltinRegistries.html#REGISTRIES)
 - ask for a codec and a network codec (which may be the same), and put those into [`DynamicRegistryManager.INFOS`](https://maven.quiltmc.org/repository/release/org/quiltmc/quilt-mappings/1.19.2+build.12/quilt-mappings-1.19.2+build.12-javadoc.jar/net/minecraft/util/registry/DynamicRegistryManager.html#INFOS)
 
 And I believe that's it.

Reply:
I think I'd prefer splitting the class into a `QuiltBuiltinRegistryHolder` and a `QuiltDynamicRegistryHolder`?

Reply:
Sure, and have a base builder class to extend, sounds fine.